classes/services/PKPSubmissionFileService.inc.php

```diff
@@ -0,0 +1,837 @@
+			if ($hasEditorialAssignment) {
+				$allowedFileStages[] = SUBMISSION_FILE_INTERNAL_REVIEW_FILE;
+			}
+		}
+
+		if (array_key_exists(WORKFLOW_STAGE_ID_EXTERNAL_REVIEW, $stageAssignments)) {
+			$hasEditorialAssignment = !empty(array_intersect($notAuthorRoles, $stageAssignments[WORKFLOW_STAGE_ID_EXTERNAL_REVIEW]));
+			// Authors can only write revision files under specific conditions
+			if ($action === SUBMISSION_FILE_ACCESS_READ || $hasEditorialAssignment) {
+				$allowedFileStages[] = SUBMISSION_FILE_REVIEW_REVISION;
+			}
+			// Authors can never access review files
+			if ($hasEditorialAssignment) {
+				$allowedFileStages[] = SUBMISSION_FILE_REVIEW_FILE;
+			}
+		}
+
+		if (array_key_exists(WORKFLOW_STAGE_ID_EDITING, $stageAssignments)
+				&& !empty(array_intersect($allowedRoles, $stageAssignments[WORKFLOW_STAGE_ID_EDITING]))) {
+			$hasEditorialAssignment = !empty(array_intersect($notAuthorRoles, $stageAssignments[WORKFLOW_STAGE_ID_EDITING]));
+			// Authors only have read access
+			if ($action === SUBMISSION_FILE_ACCESS_READ || $hasEditorialAssignment) {
+				$allowedFileStages[] = SUBMISSION_FILE_COPYEDIT;
+			}
+			if ($hasEditorialAssignment) {
+				$allowedFileStages[] = SUBMISSION_FILE_FINAL;
+			}
+		}
+
+		if (array_key_exists(WORKFLOW_STAGE_ID_PRODUCTION, $stageAssignments)
+				&& !empty(array_intersect($allowedRoles, $stageAssignments[WORKFLOW_STAGE_ID_PRODUCTION]))) {
+			$hasEditorialAssignment = !empty(array_intersect($notAuthorRoles, $stageAssignments[WORKFLOW_STAGE_ID_PRODUCTION]));
+			// Authors only have read access
+			if ($action === SUBMISSION_FILE_ACCESS_READ || $hasEditorialAssignment) {
+				$allowedFileStages[] = SUBMISSION_FILE_PROOF;
+			}
+			if ($hasEditorialAssignment) {
+				$allowedFileStages[] = SUBMISSION_FILE_PRODUCTION_READY;
+			}
+		}
+
+		HookRegistry::call('SubmissionFile::assignedFileStages', [&$allowedFileStages, $stageAssignments, $action]);
+
+		return $allowedFileStages;
+	}
+
+	/**
+	 * Get all valid file stages
+	 *
+	 * @return array
+	 */
+	public function getFileStages() {
+		import('lib.pkp.classes.submission.SubmissionFile');
+		$stages = [
+			SUBMISSION_FILE_SUBMISSION,
+			SUBMISSION_FILE_NOTE,
+			SUBMISSION_FILE_REVIEW_FILE,
+			SUBMISSION_FILE_REVIEW_ATTACHMENT,
+			SUBMISSION_FILE_FINAL,
+			SUBMISSION_FILE_COPYEDIT,
+			SUBMISSION_FILE_PROOF,
+			SUBMISSION_FILE_PRODUCTION_READY,
+			SUBMISSION_FILE_ATTACHMENT,
+			SUBMISSION_FILE_REVIEW_REVISION,
+			SUBMISSION_FILE_DEPENDENT,
+			SUBMISSION_FILE_QUERY,
+		];
+
+		HookRegistry::call('SubmissionFile::fileStages', [&$stages]);
+
+		return $stages;
+	}
+
+	/**
+	 * Get the path to a submission's file directory
+	 *
+	 * This returns the relative path from the files_dir set in the config.
+	 *
+	 * @param int $contextId
+	 * @param int $submissionId
+	 * @return string
+	 */
+	public function getSubmissionDir($contextId, $submissionId) {
+		$dirNames = Application::getFileDirectories();
+		return sprintf(
```

Comment:
Does Flysystem provide more more reliable techniques for assembling paths than string concatenation?

Reply:
I don't think so. You always just send it a `$path`. I find string concatenation easier to mess up by accident than `sprintf` but I have no strong opinions here. Happy to use whatever method we want.

Reply:
OK, press on!